### PR TITLE
Integrating class symbol support

### PR DESCRIPTION
**Pull request by andyleejordan**
WIP: Attempting to integrate https://github.com/PowerShell/PowerShellEditorServices/pull/1886 with the changes previously made in https://github.com/PowerShell/PowerShellEditorServices/pull/1917.

### CODE REVIEW

**Comment by andyleejordan** [comment]
@seeminglyscience with the integration pretty much all the changes you had to make to the unit tests I reverted. Can you explain if any of those were intentional (beyond just "making them pass")?

File: src/PowerShellEditorServices/Services/Symbols/ReferenceTable.cs [hunk 1]
```diff
@@ -19,32 +21,46 @@ internal sealed class ReferenceTable
...
     private bool IsInitialized => !_symbolReferences.IsEmpty || _isInited;
 
-    internal bool TryGetReferences(string command, out ConcurrentBag<IScriptExtent>? references)
+    internal bool TryGetReferences(string command, out ConcurrentBag<SymbolReference>? references)
     {
         EnsureInitialized();
         return _symbolReferences.TryGetValue(command, out references);
     }
 
+    // TODO: Should this be improved, or pre-sorted?
```
Comment (andyleejordan) [inline src/PowerShellEditorServices/Services/Symbols/ReferenceTable.cs]: @seeminglyscience a question for you.

File: src/PowerShellEditorServices/Services/Symbols/ReferenceTable.cs [hunk 2]
```diff
@@ -55,64 +71,146 @@ internal void EnsureInitialized()
...
+            );
 
             return AstVisitAction.Continue;
+        }
 
-            static string? GetCommandName(CommandAst commandAst)
+        // TODO: We should examine if we really want to constrain the extents to the name only. This
+        // means that highlighting only highlights the symbol name, but providing the whole extend
+        // means the whole function (or class etc.) gets highlighted, which seems to be a personal
+        // preference.
```
Comment (andyleejordan) [inline src/PowerShellEditorServices/Services/Symbols/ReferenceTable.cs]: Per https://github.com/PowerShell/PowerShellEditorServices/issues/1888 and https://github.com/PowerShell/PowerShellEditorServices/issues/1904 the current approach is valid (highlight only the name), so I need to remove this comment and we should add some tests.

File: src/PowerShellEditorServices/Services/Symbols/ReferenceTable.cs [hunk 3]
```diff
@@ -55,64 +71,146 @@ internal void EnsureInitialized()
...
+            IScriptExtent nameExtent = VisitorUtils.GetNameExtent(propertyMemberAst, false);
+            _references.AddReference(
+                SymbolType.Property,
+                VisitorUtils.GetMemberOverloadName(propertyMemberAst, false),
+                nameExtent);
+
+            return AstVisitAction.Continue;
+        }
+
+        // TODO: What else can we implement?
```
Comment (andyleejordan) [inline src/PowerShellEditorServices/Services/Symbols/ReferenceTable.cs]: Big open question here: how much else can we replace with this singular reference visitor class? Could we potentially identify when the "reference" is a definition, save that to the `SymbolReference`, and then use it later instead of searching all over again?

File: src/PowerShellEditorServices/Services/Symbols/SymbolsService.cs [hunk 4]
```diff
@@ -109,23 +110,18 @@ internal class SymbolsService
...
+            List<SymbolReference> symbols = new();
             foreach (IDocumentSymbolProvider symbolProvider in GetDocumentSymbolProviders())
             {
-                foreach (SymbolReference reference in symbolProvider.ProvideDocumentSymbols(scriptFile))
-                {
-                    reference.SourceLine = scriptFile.GetLine(reference.ScriptRegion.StartLineNumber);
-                    reference.FilePath = scriptFile.FilePath;
-                    foundOccurrences.Add(reference);
-                }
+                // TODO: Each provider needs to set the source line and filepath.
```
Comment (andyleejordan) [inline src/PowerShellEditorServices/Services/Symbols/SymbolsService.cs]: So this irked me, a lot existing logic set the file path and source line at random times instead of during the initialization of the symbol reference. Which means I think something is broken.

File: src/PowerShellEditorServices/Services/Symbols/SymbolsService.cs [hunk 5]
```diff
@@ -160,21 +156,20 @@ public List<SymbolReference> FindSymbolsInFile(ScriptFile scriptFile)
...
             ScriptFile[] referencedFiles,
-            WorkspaceService workspace,
             CancellationToken cancellationToken = default)
         {
             if (foundSymbol == null)
             {
                 return null;
             }
 
+            // TODO: Should we handle aliases at a lower level?
```
Comment (andyleejordan) [inline src/PowerShellEditorServices/Services/Symbols/SymbolsService.cs:168]: Like for real we should do this better.

File: src/PowerShellEditorServices/Services/Symbols/SymbolsService.cs [hunk 6]
```diff
@@ -227,36 +222,21 @@ static string[] GetIdentifiers(string symbolName, SymbolType symbolType, Command
...
-                        reference.FilePath = file.FilePath;
-                        symbolReferences.Add(reference);
-                    }
+                    fileReferences.AddRange(references);
 
                     await Task.Yield();
                     cancellationToken.ThrowIfCancellationRequested();
                 }
+
+                symbolReferences.AddRange(fileReferences.OrderBy(symbol => symbol.ScriptRegion.StartOffset));
```
Comment (andyleejordan) [inline src/PowerShellEditorServices/Services/Symbols/SymbolsService.cs]: So sorting has been interesting. I don't know if VS Code cares, but the tests sure do.

File: src/PowerShellEditorServices/Services/Symbols/SymbolsService.cs [hunk 7]
```diff
@@ -284,7 +264,12 @@ static string[] GetIdentifiers(string symbolName, SymbolType symbolType, Command
                 return null;
             }
 
-            return AstOperations.FindReferencesOfSymbol(file.ScriptAst, foundSymbol).ToArray();
+            if (file.References.TryGetReferences(foundSymbol.SymbolName, out ConcurrentBag<SymbolReference> references))
+            {
+                return references.OrderBy(symbol => symbol.ScriptRegion.StartOffset).ToArray();
```
Comment (andyleejordan) [inline src/PowerShellEditorServices/Services/Symbols/SymbolsService.cs]: Ditto.

File: test/PowerShellEditorServices.Test/Language/SymbolsServiceTests.cs [hunk 8]
```diff
@@ -154,9 +153,9 @@ public async Task FindsFunctionDefinitionForAlias()
         public async Task FindsReferencesOnFunction()
         {
             List<SymbolReference> referencesResult = await GetReferences(FindsReferencesOnFunctionData.SourceDetails).ConfigureAwait(true);
-            Assert.Equal(2, referencesResult.Count);
-            Assert.Equal(3, referencesResult[0].ScriptRegion.StartLineNumber);
-            Assert.Equal(2, referencesResult[0].ScriptRegion.StartColumnNumber);
+            Assert.Equal(3, referencesResult.Count);
+            Assert.Equal(1, referencesResult[0].ScriptRegion.StartLineNumber);
+            Assert.Equal(10, referencesResult[0].ScriptRegion.StartColumnNumber);
```
Comment (andyleejordan) [inline test/PowerShellEditorServices.Test/Language/SymbolsServiceTests.cs]: All these numbers essentially reverted to their values from before @seeminglyscience's cache work (they were changed in https://github.com/PowerShell/PowerShellEditorServices/pull/1917).

File: test/PowerShellEditorServices.Test/Language/SymbolsServiceTests.cs [hunk 9]
```diff
@@ -285,7 +285,7 @@ public void FindsOccurrencesOnClass()
         {
             IReadOnlyList<SymbolReference> occurrencesResult = GetOccurrences(FindsOccurrencesOnTypeSymbolsData.ClassSourceDetails);
             Assert.Equal(2, occurrencesResult.Count);
-            Assert.Equal("[SuperClass]", occurrencesResult[occurrencesResult.Count - 1].SymbolName);
+            Assert.Equal("SuperClass", occurrencesResult[occurrencesResult.Count - 1].SymbolName);
```
Comment (andyleejordan) [inline test/PowerShellEditorServices.Test/Language/SymbolsServiceTests.cs]: Except I don't know why the name for this changed.

File: test/PowerShellEditorServices.Test/Language/SymbolsServiceTests.cs [hunk 10]
```diff
@@ -312,7 +312,7 @@ public void FindsOccurrencesOnEnum()
         {
             IReadOnlyList<SymbolReference> occurrencesResult = GetOccurrences(FindsOccurrencesOnTypeSymbolsData.EnumSourceDetails);
             Assert.Equal(4, occurrencesResult.Count);
-            Assert.Equal("[MyEnum]", occurrencesResult[occurrencesResult.Count - 1].SymbolName);
+            Assert.Equal("MyEnum", occurrencesResult[occurrencesResult.Count - 1].SymbolName);
```
Comment (andyleejordan) [inline test/PowerShellEditorServices.Test/Language/SymbolsServiceTests.cs]: Ditto.

File: test/PowerShellEditorServices.Test.Shared/References/FunctionReference.ps1 [hunk 11]
```diff
@@ -0,0 +1,24 @@
+function BasicFunction {}
```
Comment (andyleejordan) [inline test/PowerShellEditorServices.Test.Shared/References/FunctionReference.ps1:1]: I had to make this a file so that the test can get a `ScriptFile`... which has the `References` field which runs the new cached `ReferenceVisitor`.

**Review by andyleejordan — COMMENTED** [review]
My own questions for you both.

**Comment by andyleejordan** [comment]
Hm ok so the sorting isn't quite right. Debugging `FindsReferencesOnTypeConstraint` and while all four references are found, they're out of order because the one used as the param type constraint has a `StartOffset` of 0.

File: src/PowerShellEditorServices/Services/Symbols/ReferenceTable.cs (hunk 3 above)
Comment (fflaten) [reply src/PowerShellEditorServices/Services/Symbols/ReferenceTable.cs]: > Could we potentially identify when the "reference" is a definition, save that to the `SymbolReference`, and then use it later instead of searching all over again?

I suggested a `IsDefinition`-prop at one point here also. https://github.com/PowerShell/PowerShellEditorServices/issues/1889#issuecomment-1215560718

That might let us replace `FindDeclarationVisitor.cs`

**Comment by fflaten** [comment]
Currently shows variable-assignments for every parameter and re-assignments inside functions/methods in outline + workspace search. Feels messy compared to stable. 

Maybe stick with current behavior for now? If so, it might require either of these:
- Sticking with `FindSymbolsVisitor` for documentsymbols and workspacesymbols handlers.
- Add another flag or bool in `SymbolReference` to control visibility in the handlers
- Include a `Parent`-backlink in the `SymbolReference` we can filter on in handler (to keep only root-level). This is easier to achieve now with a single reference visitor and is probably required to support `DocumentSymbol.children` anyway, so I'd prefer this.

File: test/PowerShellEditorServices.Test/Language/SymbolsServiceTests.cs (hunk 9 above)
Comment (fflaten) [reply test/PowerShellEditorServices.Test/Language/SymbolsServiceTests.cs:294]: Because you ordered `GetOccurrences()`. Don't see any code changed, so `TypeExpression` and `TypeConstraint` includes brackets while `TypeDefinition` (this) does not. :)

File: test/PowerShellEditorServices.Test/Language/SymbolsServiceTests.cs (hunk 10 above)
Comment (fflaten) [reply test/PowerShellEditorServices.Test/Language/SymbolsServiceTests.cs:322]: Same as above

File: src/PowerShellEditorServices/Services/Symbols/SymbolsService.cs [hunk 12]
```diff
@@ -284,7 +263,12 @@ static string[] GetIdentifiers(string symbolName, SymbolType symbolType, Command
                 return null;
             }
 
-            return AstOperations.FindReferencesOfSymbol(file.ScriptAst, foundSymbol).ToArray();
+            if (file.References.TryGetReferences(foundSymbol.SymbolName, out ConcurrentBag<SymbolReference> references))
```
Comment (fflaten) [inline src/PowerShellEditorServices/Services/Symbols/SymbolsService.cs:266]: Filter on symbol type. Will currently match ex. a similar function name and enum/class name

File: test/PowerShellEditorServices.Test/Language/SymbolsServiceTests.cs (hunk 9 above)
Comment (fflaten) [reply test/PowerShellEditorServices.Test/Language/SymbolsServiceTests.cs]: I have no idea. Maybe the tests didn't pass after all (can't remember). This seems correct though. 👍

File: test/PowerShellEditorServices.Test/Language/SymbolsServiceTests.cs [hunk 13]
```diff
@@ -256,10 +255,11 @@ public void FindsOccurrencesOnParameter()
         public async Task FindsReferencesOnCommandWithAlias()
         {
             List<SymbolReference> referencesResult = await GetReferences(FindsReferencesOnBuiltInCommandWithAliasData.SourceDetails).ConfigureAwait(true);
-            Assert.Equal(4, referencesResult.Count);
-            Assert.Equal("Get-ChildItem", referencesResult[1].SymbolName);
-            Assert.Equal("Get-ChildItem", referencesResult[2].SymbolName);
-            Assert.Equal("Get-ChildItem", referencesResult[referencesResult.Count - 1].SymbolName);
+            Assert.Collection(referencesResult,
+                (i) => Assert.Equal("Get-ChildItem", i.SymbolName),
+                (i) => Assert.Equal("gci", i.SymbolName),
```
Comment (fflaten) [inline test/PowerShellEditorServices.Test/Language/SymbolsServiceTests.cs]: 👍 Why was this changed in https://github.com/PowerShell/PowerShellEditorServices/pull/1917 in the first place? Never made sense to me, but didn't investigate.

File: src/PowerShellEditorServices/Services/Symbols/SymbolsService.cs [hunk 14]
```diff
@@ -284,7 +263,12 @@ static string[] GetIdentifiers(string symbolName, SymbolType symbolType, Command
...
+            {
+                return references;
+            }
```
Comment (fflaten) [inline src/PowerShellEditorServices/Services/Symbols/SymbolsService.cs]:
```suggestion
            if (file.References.TryGetReferences(foundSymbol.SymbolName, out ConcurrentBag<SymbolReference> references))
            {
return references.Where((r) =>
                    foundSymbol.SymbolType == r.SymbolType ||
                    (foundSymbol.SymbolType is SymbolType.Class or SymbolType.Enum or SymbolType.Type
                        && r.SymbolType is SymbolType.Class or SymbolType.Enum or SymbolType.Type)
                );
            }
```

Filter on symbol type. Will currently match ex. a function and enum/class/type with same name.

**Comment by andyleejordan** [comment]
> * Include a `Parent`-backlink in the `SymbolReference` we can filter on in handler (to keep only root-level). This is easier to achieve now with a single reference visitor and is probably required to support `DocumentSymbol.children` anyway, so I'd prefer this.

Same, that's what I was thinking.

File: test/PowerShellEditorServices.Test/Language/SymbolsServiceTests.cs (hunk 13 above)
Comment (andyleejordan) [reply test/PowerShellEditorServices.Test/Language/SymbolsServiceTests.cs]: Ditto to asking why and not being sure, but not investigating either.

File: src/PowerShellEditorServices/Services/Symbols/ReferenceTable.cs (hunk 3 above)
Comment (andyleejordan) [reply src/PowerShellEditorServices/Services/Symbols/ReferenceTable.cs]: This came along nicely!

File: src/PowerShellEditorServices/Services/Symbols/ReferenceTable.cs [hunk 15]
```diff
@@ -55,62 +71,228 @@ internal void EnsureInitialized()
...
+            _references.AddReference(
+                SymbolType.Property,
+                memberName,
+                memberExpressionAst.Member.Extent,
+                memberExpressionAst.Extent);
+
+            return AstVisitAction.Continue;
+        }
+
+        public override AstVisitAction VisitInvokeMemberExpression(InvokeMemberExpressionAst methodCallAst)
```
Comment (fflaten) [inline src/PowerShellEditorServices/Services/Symbols/ReferenceTable.cs]: Which scenario is this for? I don't think you'll be able to match it to the correct method to count it as a reference.

Comment (andyleejordan) [reply src/PowerShellEditorServices/Services/Symbols/ReferenceTable.cs]: So it turns out it works quite well! By bucketing all the overloads (of the same name) together, and returning a set of possible definitions, VS Code just does what it needs so you can do:

```powershell
class SuperClass : BaseClass {
    [string]MyClassMethod([string]$param1, $param2, [int]$param3) {
        $this.SomePropWithDefault = 'something happend'
        $param1
        return 'finished'
    }

    [string]
    MyClassMethod([MyEnum]$param1) {
        return 'hello world'
    }
    [string]MyClassMethod() {
        return 'hello world'
    }
}

$o = [SuperClass]::new()
$o.MyClassMethod()
```

and find definition on `MyClassMethod()` shows the three potential matches:
Screenshot 2023-01-20 at 5 48 12 PM

No, it's not perfect, but it's a _whole lot better_ than nothing!

Comment (fflaten) [reply src/PowerShellEditorServices/Services/Symbols/ReferenceTable.cs]: True, go to definition makes sense. 👍 You could also filter on argument count, but argument types won't work.

Comment (fflaten) [reply src/PowerShellEditorServices/Services/Symbols/ReferenceTable.cs]: To make it complete, should `::new()` return the constructors? Because that's a reserved method when static (`InvokeMemberExpressionAst.Static == true`), right?

Comment (andyleejordan) [reply src/PowerShellEditorServices/Services/Symbols/ReferenceTable.cs]: That's a good idea!
